[Blackjack - step2] Submit saragrosser's mission

saragrosser opened the pull request:
## STEP 2 BETTING SYSTEM 

In this PR, I’ve implemented all the required features for Step 2 – the Betting System in the Blackjack project. Below is an overview of the features I added, some of the challenges I faced while working on them, and the reasoning behind a few design decisions I made along the way.

### **Implemented Features**
- Bet Input: Each player is now prompted to place a bet before the game starts.

- Input Validation: Bets must be a positive number and a multiple of 1,000. If not, the player is asked again until valid input is given.

- Blackjack Bonus: If a player hits Blackjack with the initial two cards, they receive 1.5x their bet.

- Blackjack Draw Rule: If both the player and the dealer have Blackjack, the player’s bet is returned as a draw.

- Earnings Logic: Players earn or lose money based on the outcome of the game (WIN = +bet, LOSE = –bet, etc.).

- Dealer Earnings: The dealer's earnings are now calculated as the inverse of the sum of all player earnings.

- Final Output: The final results section displays the result of each player and their earnings, along with the dealer's.

### **Refactoring & Design Decisions**
I introduced a BettingLogic object to better separate concerns and follow the Single Responsibility Principle. At first, I handled earnings calculation in the controller, but I realized it would be cleaner to isolate that logic in its own component.

I also kept some logic in GameLogic, such as result evaluation and Blackjack handling, to reduce clutter in the controller and keep responsibilities clear.

I considered removing GameLogic and placing everything directly in the controller, but it started getting harder to maintain, so I kept the helper logic separated.

### **Challenges & Fixes**
- Blackjack Bonus & Draw Handling: Initially, players with Blackjack weren’t receiving the proper bonus or draw logic. To fix this, I added extra parameters (isPlayerBlackjack and isDealerBlackjack) to the getGameResult() method and updated the result logic accordingly.

- Dealer Earnings Always 0: At first, the dealer’s earnings were stuck at 0. I realized I wasn’t subtracting player earnings properly and fixed it by summing up the negatives of all player earnings.

- Bet Input Validation Loop: I wanted to make sure players could only bet in multiples of 1,000. I had to write a retry mechanism that gracefully asks the player again if the input is invalid, without crashing the app.

### **Testing**
I updated GameLogicTest.kt to reflect the new rules, especially around Blackjack wins and draws.

I tested several scenarios manually to confirm all payouts are correct and edge cases (like multiple Blackjacks or busts) behave as expected.

Let me know how this approach was, I'm  open to all feedback!!

Thanks for taking the time to review this! 😊

Best,
Sara

saragrosser committed:
docs(readme): add betting system feature checklist for step 2

saragrosser committed:
feat(betting): prompt each player to input their betting amount

saragrosser committed:
test(betting): add unit test for BettingInfo data class

saragrosser committed:
feat(betting): calculate and display earnings based on game results and bets

test(betting): add unit tests for result and earnings calculation

saragrosser committed:
style: format code with ktlint

saragrosser committed:
feat(BettingLogic): add BettingLogic.kt with corresponding BettingLogicTest.kt file to manage player's bets in the game

saragrosser committed:
refactor(game-controller): remove unnecessary logic to store bets and display final earnings

saragrosser committed:
refactor(input-view): add bet constraint to accept only multiples of 1000

saragrosser committed:
docs(readme): update checklist of betting system features

saragrosser committed:
fix(GameLogic): treat both dealer and player Blackjack as draw

saragrosser committed:
fix(GameLogic): treat both dealer and player Blackjack as draw

saragrosser committed:
style: formatting with ktlintFormat

saragrosser committed:
refactor(game-controller): extracted logic from run method into private methods to make code more readable

saragrosser committed:
style: run ktlintFormat and change directory name to have a more intuitive name instead of controller

Hyeon9mak commented on README.md:
```diff
@@ -26,3 +26,13 @@ Includes card dealing, drawing logic, and win/loss/draw result handling.
 - [x] Display final hands and scores
 - [x] Determine and print winners using `GameLogic.getGameResult()`
+
+### Betting System
+- [x] Prompt each player to enter a betting amount at the start of the game
+- [x] Store and associate each player's bet with their name
+- [x] Calculate 1.5x payout for Blackjack (21 with initial two cards)
+- [x] Handle full loss of bet if player busts
+- [x] Handle dealer bust: remaining players win their bets
+- [x] Compare scores between dealer and players to calculate win/loss
+- [x] Display final earnings for each player and the dealer
+
```
😉 👍

Hyeon9mak commented on src/main/kotlin/blackjack/blackjackgame/BettingLogic.kt:
```diff
@@ -0,0 +1,18 @@
+package blackjack.blackjackgame
+
+import blackjack.model.GameResult
+
+object BettingLogic {
+    fun calculateEarnings(
```
https://github.com/herotechcourse/kotlin-blackjack/pull/11#discussion_r2204968757

I think `Logic` is common name too. 😢

saragrosser replied:
Thank you! I renamed it to "BetEarningsCalculator" to reflect better the responsibility :)

Hyeon9mak commented on src/main/kotlin/blackjack/blackjackgame/BettingLogic.kt:
```diff
@@ -0,0 +1,18 @@
+            result == GameResult.LOSE -> -bet
+            else -> 0
+        }
+    }
```
I think the `GameResult` can get calculate bet logic.

```kotlin
enum class GameResult(
    private val enarning: Double,
) {
    BLACKJACK_WIN(1.5),
    WIN(1.0),
    // ...
}
```

enum class is not a just constants collection.
it can also serve as a group of logic that represents rules. 😉

saragrosser replied:
Hey Hyeon thank you so much this was super helpful and actually I ended up refactoring and deleting BettinLogic file because it can now be handle inside GameResult enum class! 
Thank you!

Hyeon9mak commented on src/main/kotlin/blackjack/model/GameResult.kt:
```diff
@@ -3,5 +3,5 @@ package blackjack.model
-}
\ No newline at end of file
+    DRAW("Draw"),
+}
```
https://github.com/herotechcourse/kotlin-blackjack/pull/11#discussion_r2204941171

saragrosser replied:
Done :)

Hyeon9mak commented on src/main/kotlin/blackjack/view/SuitView.kt:
```diff
@@ -6,7 +6,8 @@ enum class SuitView(val suit: Suit, val symbol: String) {
     HEART(Suit.HEART, "♥"),
     DIAMOND(Suit.DIAMOND, "♦"),
     CLUB(Suit.CLUB, "♣"),
-    SPADE(Suit.SPADE, "♠");
+    SPADE(Suit.SPADE, "♠"),
```
https://github.com/herotechcourse/kotlin-blackjack/pull/11#discussion_r2204939341

saragrosser replied:
Done I commented on the step 1 PR feedback message

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -4,7 +4,6 @@ import blackjack.view.InputView
-
     private val deck = Deck()
 
     fun giveTwoCardsTo(participants: List<Participant>) {
```
https://github.com/herotechcourse/kotlin-blackjack/pull/11#discussion_r2204949837

saragrosser replied:
Done I commented on the step 1 PR feedback message

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -4,7 +4,6 @@ import blackjack.view.InputView
 import blackjack.view.OutputView
```
https://github.com/herotechcourse/kotlin-blackjack/pull/11#discussion_r2204946072

saragrosser replied:
Done :)

Hyeon9mak commented on src/main/kotlin/blackjack/blackjackgame/GameLogic.kt:
```diff
@@ -0,0 +1,38 @@
+    fun handlePlayerTurn(
+        dealer: Dealer,
+        player: Player,
+    ) {
+        if (player.isBlackJack()) {
+            println("${player.name} has Blackjack!")
+            return
+        }
```
nooooooo! u said u already learn about MVC pattern! 👿

saragrosser replied:
sorrryyyyyyyy!!!! I fixed it now!

Hyeon9mak commented on src/main/kotlin/blackjack/blackjackgame/BlackjackGame.kt:
```diff
@@ -0,0 +1,97 @@
+        OutputView.displayDealerStatus(dealer)
+    }
+
+    private fun displayGameResults(
+        players: List<Player>,
+        dealer: Dealer,
+    ) {
+        val dealerScore = dealer.getScore()
+        val isDealerBusted = dealer.isBusted()
+        val isDealerBlackjack = dealer.isBlackJack()
+
+        var dealerWins = 0
```
> Don't use getter. Send(Ask) message to Object.

I think u can ask dealer or player about their win/lose!

```kotlin
dealer.winOrLose(player)
players.winOrLose(dealer)
```

saragrosser replied:
Thank you for your patience! It really helped me understand the importance of respecting object responsibility and avoiding getter-heavy logic.

I moved the game result evaluation into Dealer.judge(player) so each object can now take care of its own logic. I also removed the getGameResult() method and updated the related test cases using simple test doubles (and marked methods as open for testability).
Finally, I fixed the MVC violation by moving the Blackjack message to OutputView. hehe

Hyeon9mak commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -6,6 +6,18 @@ object InputView {
         return input.split(",").map { it.trim() }.filter { it.isNotEmpty() }
     }
+
+    fun getBettingAmount(playerName: String): Int {
+        println("Enter $playerName’s betting amount (multiple of 1000):")
+        val input = readLine()?.toIntOrNull()
+        return if (input != null && input % 1000 == 0 && input > 0) {
+            input
+        } else {
+            println("Invalid amount. Please enter a positive number that is a multiple of 1000.")
```
It's hard to generate test code. right?

I think u can create the `Bet` value class, and append validation, and generate test code more easier!

saragrosser replied:
I introduced a Bet value class to encapsulate the validation logic. This helped remove the inline validation from InputView and made the betting logic much easier to reason about and test.

Also updated the Player class and related usages to work with Bet directly instead of a raw Int.

Really appreciate this kind of feedback it's helping me build better habits around modeling the domain clearly! 🙏

Hyeon9mak reviewed (changes requested):
Hi Sari! step2 wasn’t too challenging, was it? 😊

You did a good job with this step overall,
but I noticed that some feedback from the previous stage hasn’t been fully addressed, which is a bit disappointing... 😢 

I’ve added links to the relevant parts of the earlier feedback — it would be great if you could reflect on them or share your own thoughts on those points.

I’ve also left a few additional minor comments.
If you have any questions, feel free to ask!

saragrosser committed:
refactor: change class names to best reflect responsabilities

saragrosser committed:
refactor: Update GameResult.kt file to handle the calculations for bet logic

-Delete BetEarningsCalculator.kt file beacuase it is not needed anymore

saragrosser committed:
refactor: use ktlinformatting to improve the format

saragrosser committed:
refactor(core): move game result logic to Dealer and enforce MVC

- Moved getGameResult() logic from TurnManager to Dealer.judge(player)
- Replaced getter-based comparisons with encapsulated method calls
- Removed direct println() call from TurnManager to respect MVC
- Moved blackjack message to OutputView
- Renamed GameRoundManager to TurnManager for clearer responsibility
- Deleted unused getGameResult() function
- Deleted outdated GameRoundManagerTest

saragrosser committed:
refactor(mvc): remove view dependency from Dealer and delegate turn control to TurnManager

- Moved askPlayerDraw and playDealerTurn logic out of Dealer
- Dealer no longer depends on InputView or OutputView
- Added TurnManager.handleDealerTurn to coordinate dealer drawing with view updates
- Passed "Dealer" name explicitly to Participant to preserve display needs

saragrosser committed:
refactor(view): make suit and rank private in SuitView and RankView enums

saragrosser committed:
refactor(view): extract GameResult labels to GameResultView enum

- Removed label property from GameResult model to follow MVC separation
- Created GameResultView enum for UI representation of GameResult
- Updated BlackjackGame to use GameResultView.from(result).label
- Preserved display logic in OutputView while keeping GameResult domain-focused

saragrosser committed:
refactor(domain): introduce Bet value class for betting validation

- Created Bet value class to encapsulate and validate betting rules
- Replaced raw Int bets with strongly typed Bet objects
- Updated InputView to return Bet and removed inline validation
- Updated Player class and usage to accept Bet type instead of Int

Hyeon9mak commented on src/main/kotlin/blackjack/model/GameResult.kt:
```diff
@@ -1,7 +1,15 @@
+    BLACKJACK_WIN(1.5),
+    WIN(1.0),
+    DRAW(0.0),
+    LOSE(-1.0),
```
much better 😉 👍 
also u can try like this

```suggestion
enum class GameResult(val calculateEarnings: (Int) -> Int) {
    BLACKJACK_WIN({ bet -> (bet * 1.5).toInt() }),
    WIN({ bet -> bet }),
    DRAW({ _ -> 0 }),
    LOSE({ bet -> -bet });
}
```

Hyeon9mak replied:
And, `GameResult` also can use the `BetAmount` 😉

saragrosser replied:
Thank you I took your suggestion, I replaced the multiplier with function-based strategy in GameResult, and also updated it to use the Bet type directly :)

Hyeon9mak commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -38,15 +37,34 @@ object OutputView {
+        players: List<Player>,
+        dealerEarnings: Int,
+    ) {
+        println("\n## Final Earnings")
+        println("Dealer: $dealerEarnings")
+        players.forEach { println("${it.name}: ${it.earnings}") }
```
> I've now removed all references to InputView and OutputView from Dealer and moved the related logic into TurnManager. The model no longer depends on any view layer, and the controller fully coordinates interactions between model and view.

niiice! now when we need change dealer presentation to image, we just check view layer only!

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -19,28 +15,28 @@ class Dealer : Participant("Dealer") {
-            OutputView.displayDealerDraw()
-            giveCardTo(this)
-            OutputView.displayDealerStatus(this)
+    fun judge(player: Player): GameResult {
+        val playerScore = player.getScore()
+        val isPlayerBusted = player.isBusted()
+        val isPlayerBlackjack = player.isBlackJack()
+
+        val dealerScore = this.getScore()
+        val isDealerBusted = this.isBusted()
+        val isDealerBlackjack = this.isBlackJack()
+
+        return when {
+            isPlayerBusted -> GameResult.LOSE
+            isDealerBusted -> GameResult.WIN
+            isPlayerBlackjack && isDealerBlackjack -> GameResult.DRAW
+            isPlayerBlackjack -> GameResult.WIN
+            isDealerBlackjack -> GameResult.LOSE
+            playerScore > dealerScore -> GameResult.WIN
+            playerScore < dealerScore -> GameResult.LOSE
+            else -> GameResult.DRAW
```
`judge` nice naming! 💯 👍 
it makes much easier create test code 👍

saragrosser replied:
Yes it make much more sense, thank you!

Hyeon9mak commented on src/main/kotlin/blackjack/model/BetAmount.kt:
```diff
@@ -0,0 +1,13 @@
+            }
+            return Bet(amount)
+        }
+    }
```
If u move the validation logic into the `init` block, you'll be able to enforce validation even when an instance is created using the primary constructor. 😄

saragrosser replied:
Thanks for the suggestion, I added the validation logic to the init block so that Bet is now always validated regardless of how it's instantiated.

Hyeon9mak commented on src/main/kotlin/blackjack/model/Participant.kt:
```diff
@@ -1,15 +1,17 @@
+    open fun isBusted() = hand.isBusted()
+
+    open fun isBlackJack() = hand.isBlackjack(getScore())
 
```
i think, `name` is only `Player`'s property 😄 
how do u think?

saragrosser replied:
yes you are totally right, it doesnt make sense to pass name for the dealer. I have  now removed the `name` property from the Participant class :)

Hyeon9mak commented on src/main/kotlin/blackjack/blackjackgame/TurnManager.kt:
```diff
@@ -0,0 +1,37 @@
+            } else {
+                break
+            }
+        }
```
how about ask to player? `player.isBusted()`
like `isBlackJack()` 😄

saragrosser replied:
I updated TurnManager to use `player.isBusted()` instead of checking the raw score directly. It's now more readable and better aligned with our domain design just like with isBlackJack()

Appreciate all the great feedback and lessons from this PR 🙏😊

Hyeon9mak reviewed (changes requested):
Hi Sara, I'm so sorry for late 🙏 

You've reflected the feedback well. great job!
I also checked the comments I left on your step 1 PR. 😄 

I left a few more small suggestions that could help polish things up just a bit.

You're almost there! Just a little more to go.
If you have any questions, feel free to ask anytime.

saragrosser committed:
refactor:  use function strategy and Bet type in GameResult earnings logic

saragrosser committed:
refactor(domain): enforce Bet validation using init block

saragrosser committed:
refactor(model): remove name from Participant and restrict to Player only

saragrosser committed:
refactor(turn): use isBusted instead of raw score check in TurnManager

- Replaced `getScore() >= 21` with domain method `isBusted()`

Hyeon9mak commented on src/test/kotlin/blackjack/model/GameResultTest.kt:
```diff
@@ -0,0 +1,31 @@
+
+class GameResultTest {
+
+    @Test
+    fun `should calculate earnings for blackjack win`() {
+        val earnings = GameResult.BLACKJACK_WIN.earningsFrom(Bet.from(10000))
+        assertEquals(15000, earnings)
+    }
+
+    @Test
+    fun `should calculate earnings for regular win`() {
+        val earnings = GameResult.WIN.earningsFrom(Bet.from(10000))
+        assertEquals(10000, earnings)
+    }
+
+    @Test
+    fun `should calculate earnings for draw`() {
+        val earnings = GameResult.DRAW.earningsFrom(Bet.from(10000))
```
💯 👍

Hyeon9mak reviewed (approved):
Hi Sara!
You did a great job all the way to the end—well done!

I'm sure you've spent a lot of time thinking about how to write maintainable code.
I believe you're now well prepared to tackle the remaining missions.

I hope you enjoy the rest of the Spring mission as well!
And if you have any questions, feel free to ask anytime. 😄
